### DISCUSSION TITLE
High Azure cloud bill due to large number of API requests - best way to debug?

### DISCUSSION THREAD

**Discussion by Norbo11** [minio/minio#12188] [Q&A] [answered]
We are trying to better understand the storage portion of our Azure cloud bill. 

It appears that most of our bill comes from Azure's pricing of individual API requests that are sent against the storage account. The pricing breakdown is available at https://azure.microsoft.com/en-gb/pricing/details/storage/blobs/, e.g. £0.0373 for every 10,000 list/read container operations.

All of our systems talk to Azure blob storage via the MinIO Azure gateway. We have collected various metrics from MinIO on the amount of S3 operations that happen against our single MinIO instance. However, we are seeing a massive disconnect between how many operations we see in MinIO vs what Azure reports - we seem to send an order of magnitude more operations to Azure than what MinIO stats report. 

I did some digging into the Azure gateway code: https://github.com/minio/minio/blob/cbfdf97abf9fc05183ef65f41c692ae67ca5b721/cmd/gateway/azure/gateway-azure.go

From the above, it does seem like certain operations in that file might send multiple requests to Azure. For example, the `PutObject` function calls  `azblob.UploadStreamToBlockBlob`, `blobURL.GetProperties`, `blobURL.SetMetadata`, and `a.GetObjectInfo`, which would imply any `PutObject` request sent to the gateway translates to 4 different API calls on the Azure side.

Ideally we would love to use something like `mc admin trace` to view these operations, but this command only reports the Minio requests, and doesn't tell us how those requests translate to Azure operations. Is there anything we can do, e.g. a debug flag or a logging option that we might be able to enable on the MinIO Azure gateway, which would reveal exactly the API calls that MinIO makes to Azure? 

Ideally I would love to say something like `mc admin trace --azure` and instead of getting 
```
12:25:57.612 [200 OK] s3.PutObject our.domain.com/some_file 127.0.0.1        53.596ms     ↑ 265 B ↓ 64 KiB
12:25:57.669 [206 Partial Content] s3.GetObject our.domain.com/some_file 127.0.0.1        28.958ms     ↑ 265 B ↓ 187 KiB
12:25:57.651 [206 Partial Content] s3.GetObject our.domain.com/some_file 127.0.0.1        61.082ms     ↑ 265 B ↓ 2.6 MiB
```
I would get something like:
```
12:25:57.612 [200 OK] azure.PutBlob our.domain.com/some_file
12:25:57.612 [200 OK] azure.GetBlobProperties our.domain.com/some_file
12:25:57.612 [200 OK] azure.SetBlobMetadata our.domain.com/some_file
12:25:57.612 [200 OK] azure.GetObjectInfo our.domain.com/some_file
12:25:57.669 [206 Partial Content] azure.GetBlob our.domain.com/some_file
12:25:57.651 [206 Partial Content] azure.GetBlob our.domain.com/some_file
```

I can see that the Gateway uses the `azure-storage-blob-go` Go package to make these requests, but I'm not familiar enough with Go to figure out whether there's an easy way to toggle a debug mode using an environment variable?

Many thanks!

**Comment by klauspost** [comment, marked as answer]
As a design proposal I am against mixing concepts. `mc admin trace` shows traces of requests being made *to* MinIO servers, not from the servers. That however isn't really important.

I can see the value of tracing outgoing requests. This is not something we currently have on our design roadmap, since 'gateway' is mainly in feature freeze.

I don't know if it is feasible, but for a short time solution you could look into routing your external requests through a proxy that can log them.
